### PR TITLE
 Provide links to locally built documentation for `experimental/externalDocs`

### PR DESCRIPTION
**Pull request by Ddystopia**
This pull request addresses issue #12867, which requested the ability to provide links to locally built documentation when using the "Open docs for symbol" feature. Previously, rust-analyzer always used docs.rs for this purpose. With these changes, the feature will provide both web (docs.rs) and local documentation links without verifying their existence.

Changes in this PR:

   - Added support for local documentation links alongside web documentation links.
   - Added `target_dir` path argument for external_docs and other related methods.
   - Added `sysroot` argument for external_docs.
   - Added `target_directory` path to `CargoWorkspace`.

API Changes:
   
   - Added an experimental client capability `{ "localDocs": boolean }`. If this capability is set, the `Open External Documentation` request returned from the server will include both web and local documentation links in the `ExternalDocsResponse` object.

Here's the `ExternalDocsResponse` interface:

```typescript
interface ExternalDocsResponse {
    web?: string;
    local?: string;
}
```

By providing links to both web-based and locally built documentation, this update improves the developer experience for those using different versions of crates, git dependencies, or local crates not available on docs.rs. Rust-analyzer will now provide both web (docs.rs) and local documentation links, leaving it to the client to open the desired link. Please note that this update does not perform any checks to ensure the validity of the provided links.

### CODE REVIEW

**Comment by nemethf** [comment]
Can this be implemented in a backward compatible way?  I.e., clients conforming to the current  extension might fail in the future because the type of the return value has changed.  (One possibility to overcome this issue is make rust-analyzer check for a client capability like, for example, `"externalDocs": {"version":  2}`.)

Also, does the local documentation count as "external documentation"? 

Thank you.

**Comment by Ddystopia** [comment]
> Can this be implemented in a backward compatible way? I.e., clients conforming to the current extension might fail in the future because the type of the return value has changed. (One possibility to overcome this issue is make rust-analyzer check for a client capability like, for example, `"externalDocs": {"version": 2}`.)

As far as I know rust-analyzer does not provide a stable API interface so breakage is fine. 
Also, there is #4604 for this purpose. As stated:
> This issue exists to announce changes which might affect rust-analyzer plugins for different editors.
> If you maintain such a plugin, consider subscribing to this issue.


> Also, does the local documentation count as "external documentation"?

I assume that "external documentation" is documentation outside of the editor.

Maybe, it would be better to create something like `experimental/localDocs`?

Thank you for your comment.

File: crates/rust-analyzer/src/handlers.rs [hunk 1]
```diff
@@ -1560,13 +1560,22 @@ pub(crate) fn handle_semantic_tokens_range(
...
-) -> Result<Option<lsp_types::Url>> {
+) -> Result<(Option<lsp_types::Url>, Option<lsp_types::Url>)> {
     let _p = profile::span("handle_open_docs");
     let position = from_proto::file_position(&snap, params)?;
 
-    let remote = snap.analysis.external_docs(position)?;
+    let cargo = match snap.workspaces.get(0) {
+        Some(ProjectWorkspace::Cargo { cargo, .. }) => Some(cargo),
+        _ => None,
+    };
```
Comment (Veykril) [inline crates/rust-analyzer/src/handlers.rs:1570]: You can fetch the correct workspace similar to how its done here, https://github.com/rust-lang/rust-analyzer/blob/968850d7bf9916b4d2a6492406a10fd562120b48/crates/rust-analyzer/src/cargo_target_spec.rs#L120-L128

File: crates/ide/src/doc_links.rs [hunk 2]
```diff
@@ -308,19 +322,38 @@ fn broken_link_clone_cb(link: BrokenLink<'_>) -> Option<(CowStr<'_>, CowStr<'_>)
...
+        local_url: local_url.map(|it| it.into()),
+    };
+
+    fn join_url(base_url: Option<Url>, path: &str) -> Option<Url> {
+        base_url.and_then(|url| url.join(path).ok())
+    }
+
+    fn set_fragment_for_url(url: Option<&mut Url>, frag: Option<&str>) {
+        url.map(|url| url.set_fragment(frag));
+    }
```
Comment (Veykril) [inline crates/ide/src/doc_links.rs:356]: nit: let's make these two functions closures (there is an assist for that) and stick them at the top of the function, that way its less noisy (and we can get rid of the trailing return)

File: crates/ide/src/doc_links.rs [hunk 3]
```diff
@@ -447,16 +489,14 @@ fn get_doc_base_url(db: &RootDatabase, def: Definition) -> Option<Url> {
             | LangCrateOrigin::ProcMacro
             | LangCrateOrigin::Std
             | LangCrateOrigin::Test),
-        ) => {
-            format!("https://doc.rust-lang.org/{channel}/{origin}")
-        }
-        CrateOrigin::Lang(_) => return None,
+        ) => (Some(format!("https://doc.rust-lang.org/{channel}/{origin}")), None),
```
Comment (Veykril) [inline crates/ide/src/doc_links.rs:492]: I believe the local docs come pre-installed right? We should be able to point this to `{sysroot}/share/doc/rust/{origin}/index.html`, we'd need to thread the sysroot base path down here as well then.

File: crates/ide/src/doc_links.rs [hunk 4]
```diff
@@ -483,10 +524,23 @@ fn get_doc_base_url(db: &RootDatabase, def: Definition) -> Option<Url> {
...
+    let local_base = local_base.and_then(|it| it.join(&format!("{display_name}/")).ok());
+
+    return (web_base, local_base);
+
+    fn create_url_from_os_str(path: &OsStr) -> Option<Url> {
+        let mut with_prefix = OsStr::new("file:///").to_os_string();
+        with_prefix.push(path);
+        with_prefix.push("/");
+        with_prefix.to_str().and_then(|s| Url::parse(s).ok())
+    }
```
Comment (Veykril) [inline crates/ide/src/doc_links.rs:543]: again, lets make this a closure and put it in front of its usage (or inline it directly)

**Comment by nemethf** [comment]
> As far as I know, rust-analyzer does not provide a stable API interface, so breakage is fine. Also, there is #4604 for this purpose. As stated:
> 
> > This issue exists to announce changes which might affect rust-analyzer plugins for different editors.
> > If you maintain such a plugin, consider subscribing to this issue.

Maybe you're right.  But to me, this text only says how plugin-maintainers can keep up with the changes of the lsp-extensions and says nothing about a possible breakage.

> > Also, does the local documentation count as "external documentation"?
> 
> I assume that "external documentation" is documentation outside of the editor.
> 
> Maybe, it would be better to create something like `experimental/localDocs`?

I don't know, but I start to wonder why it is necessary to introduce protocol extensions for this feature.  The LSP client can send a standard compliant  `workspace/executeCommand`  request and the server can send a `window/showDocument` request right after receiving the executeCommand.  The name of the command and its arguments and return value still need to be documented, though.

File: crates/rust-analyzer/src/lsp_ext.rs [hunk 5]
```diff
@@ -474,7 +474,7 @@ pub enum ExternalDocs {}
 
 impl Request for ExternalDocs {
     type Params = lsp_types::TextDocumentPositionParams;
-    type Result = Option<lsp_types::Url>;
+    type Result = (Option<lsp_types::Url>, Option<lsp_types::Url>);
```
[the remark this answers is not in the document]
Comment (Ddystopia) [reply crates/rust-analyzer/src/lsp_ext.rs:477]: Should capabilities look like that?
```diff
diff --git a/crates/rust-analyzer/src/caps.rs b/crates/rust-analyzer/src/caps.rs
index 3628670ac..c6c5d99b7 100644
--- a/crates/rust-analyzer/src/caps.rs
+++ b/crates/rust-analyzer/src/caps.rs
@@ -141,6 +141,7 @@ pub fn server_capabilities(config: &Config) -> ServerCapabilities {
         inline_value_provider: None,
         experimental: Some(json!({
             "externalDocs": true,
+            "localDocs": false,
             "hoverRange": true,
             "joinLines": true,
             "matchingBrace": true,
```
Or how to set that? Thank you.

File: crates/ide/src/doc_links.rs [hunk 6]
```diff
@@ -136,21 +148,23 @@ pub(crate) fn external_docs(
     let node = token.parent()?;
     let definition = match_ast! {
         match node {
-            ast::NameRef(name_ref) => match NameRefClass::classify(sema, &name_ref)? {
```
Comment (Veykril) [inline crates/ide/src/doc_links.rs:139]: We can bring back the `?` operator here now

File: crates/ide/src/doc_links.rs [hunk 7]
```diff
@@ -308,19 +322,35 @@ fn broken_link_clone_cb(link: BrokenLink<'_>) -> Option<(CowStr<'_>, CowStr<'_>)
...
+    local_url = join_url(local_url, &file);
+
+    web_url.as_mut().map(|url| url.set_fragment(frag.as_deref()));
+    local_url.as_mut().map(|url| url.set_fragment(frag.as_deref()));
 
-    Some(url.into())
+    return DocumentationLinks {
+        web_url: web_url.map(|it| it.into()),
+        local_url: local_url.map(|it| it.into()),
+    };
```
Comment (Veykril) [inline crates/ide/src/doc_links.rs:353]:
```suggestion
    DocumentationLinks {
        web_url: web_url.map(|it| it.into()),
        local_url: local_url.map(|it| it.into()),
    }
```

**Review by Veykril — APPROVED** [review]

**Comment by Veykril** [comment]
> I don't know, but I start to wonder why it is necessary to introduce protocol extensions for this feature. The LSP client can send a standard compliant  `workspace/executeCommand`  request and the server can send a `window/showDocument` request right after receiving the executeCommand.  The name of the command and its arguments and return value still need to be documented, though.  

How is using `workspace/executeCommand` any different? The client still needs to know about the command there that the server supports. And the server replying with `window/showDocument` would limit in what the client can do, the can't decide to open things in a browser if it wants to for example.

**Comment by nemethf** [comment]
> > I don't know, but I start to wonder why it is necessary to introduce protocol extensions for this feature. The LSP client can send a standard compliant  `workspace/executeCommand`  request and the server can send a `window/showDocument` request right after receiving the executeCommand.  The name of the command and its arguments and return value still need to be documented, though.
> 
> How is using `workspace/executeCommand` any different? The client still needs to know about the command there that the server supports. And the server replying with `window/showDocument` would limit in what the client can do, the can't decide to open things in a browser if it wants to for example.

As the showDocument request supports an "external" flag, the server can implement the logic of determining how the client should open the documentation.  This way the client-side code can be smaller, which is a desirable goal, I think. 

At any rate, I'm happy with the merged feature.  Thanks @Ddystopia and @Veykril!

**Pull request merged**